src/IO/DataImporter/DataFileReaderActions.hpp

```diff
@@ -0,0 +1,219 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <cstddef>
+#include <iterator>
+#include <unordered_map>
+
+#include "DataStructures/DataBox/DataBox.hpp"
+#include "DataStructures/DataBox/DataBoxTag.hpp"
+#include "DataStructures/Index.hpp"
+#include "DataStructures/Tensor/TensorData.hpp"
+#include "ErrorHandling/Error.hpp"
+#include "IO/DataImporter/Tags.hpp"
+#include "IO/H5/AccessType.hpp"
+#include "IO/H5/File.hpp"
+#include "IO/H5/VolumeData.hpp"
+#include "IO/Observer/ArrayComponentId.hpp"
+#include "Parallel/ConstGlobalCache.hpp"
+#include "Parallel/Info.hpp"
+#include "Parallel/Invoke.hpp"
+#include "Utilities/Algorithm.hpp"
+#include "Utilities/Requires.hpp"
+#include "Utilities/TaggedTuple.hpp"
+
+namespace importer {
+/// Actions related to the data importer
+namespace Actions {
+
+/*!
+ * \brief Invoked on the `importer::DataFileReader` component to store the
+ * registered data.
+ *
+ * The `importer::Actions::RegisterWithImporter` action, which is performed on
+ * each element of an array parallel component, invokes this action on the
+ * `importer::DataFileReader` component.
+ */
+struct RegisterElementWithSelf {
+  template <
+      typename ParallelComponent, typename DbTagsList, typename Metavariables,
+      typename ArrayIndex, typename DataBox = db::DataBox<DbTagsList>,
+      Requires<db::tag_is_retrievable_v<Tags::RegisteredElements, DataBox>> =
+          nullptr>
+  static void apply(db::DataBox<DbTagsList>& box,
+                    const Parallel::ConstGlobalCache<Metavariables>& /*cache*/,
+                    const ArrayIndex& /*array_index*/,
+                    const observers::ArrayComponentId& array_component_id,
+                    const std::string& grid_name) noexcept {
```

Comment:
Newbie question: so grid_name is the name of a specific element? Are you storing their names as strings because you'll need these strings when reading volume data from an h5 file?

Reply:
Yes, the `grid_name` is the string that identifies the element in the H5 file. We currently use the ElementId for this in the observers, so the importer does the same.